File: src/main/java/work/nekow/particledrawing/util/ParticleUtils.java
package work.nekow.particledrawing.util;

import net.minecraft.resources.Identifier;
import net.minecraft.server.level.ServerLevel;

import java.util.UUID;

/**
 * Utility methods for the particle drawing system.
 */
public final class ParticleUtils {

    private ParticleUtils() {}

    /**
     * Generate a stable UUID from a dimension's ResourceLocation.
     * This ensures the same dimension always gets the same engine instance.
     */
    public static UUID dimensionUUID(ServerLevel level) {
        return dimensionUUID(level.dimension().identifier());
    }

    public static UUID dimensionUUID(Identifier location) {
        long hash = location.toString().hashCode();
        if (hash < 0) {
            return new UUID(0, Math.abs(hash));
        }
        return new UUID(hash, 0);
    }
}

File: src/main/kotlin/work/nekow/particledrawing/util/ParticleUtils.kt
package work.nekow.particledrawing.util

import net.minecraft.resources.Identifier
import net.minecraft.server.level.ServerLevel
import java.util.UUID

/**
 * 粒子相关的工具方法。
 */
object ParticleUtils {

    /**
     * 根据服务端世界生成维度 UUID。
     * @param level 服务端世界实例
     * @return 维度对应的 UUID
     */
    fun dimensionUUID(level: ServerLevel): UUID = dimensionUUID(level.dimension().identifier())

    /**
     * 根据资源标识符生成确定性 UUID（基于名称哈希）。
     * @param location 维度的资源标识符
     * @return 维度对应的 UUID
     */
    fun dimensionUUID(location: Identifier): UUID =
        UUID.nameUUIDFromBytes(location.toString().encodeToByteArray())
}
